File: test/adc_filter_test.c
#include <assert.h>
#include <stdint.h>
#include <stdio.h>

#include "src/generic/adc_filter.h"

int
main(int argc, char **argv)
{
    struct adc_filter f;
    struct adc_filter_config config = {
        .input_div = 2, .osr = 4, .report_div = 3, .shift = 2,
    };
    assert(!adc_filter_configure(&f, &config));
    struct adc_filter_summary result;
    uint32_t filtered_value = 0;
    uint8_t filtered_ready = 0;
    for (uint32_t i = 0; i < 24; i++) {
        int ready = adc_filter_push_ex(&f, i * 3, i, &result,
                                       &filtered_value, &filtered_ready);
        if (i == 6)
            assert(filtered_ready && filtered_value == 9);
        if (i != 22)
            assert(!ready);
        else {
            assert(ready);
            // Accepted groups are [0,6,12,18], [24,30,36,42], and
            // [48,54,60,66], rounded after a two-bit shift.
            assert(result.count == 3);
            assert(result.minimum == 9 && result.maximum == 57);
            assert(result.sum == 99);
            assert(result.first_scan == 6 && result.last_scan == 22);
            assert(!result.flags);
        }
    }

    adc_filter_reset(&f, 1);
    for (uint32_t i = 0; i < 22; i++)
        assert(!adc_filter_push(&f, 100, i, &result));
    assert(adc_filter_push(&f, 100, 22, &result));
    assert(result.flags == ADC_FILTER_FLAG_DISCONTINUITY);

    config = (struct adc_filter_config) {
        .input_div = 1, .osr = 2, .report_div = 3, .shift = 1,
        .summary_mode = ADC_FILTER_SUMMARY_LATEST,
    };
    assert(!adc_filter_configure(&f, &config));
    const uint16_t latest_values[] = {10, 20, 30, 40, 50, 60};
    for (uint32_t i = 0; i < 6; i++) {
        int ready = adc_filter_push(&f, latest_values[i], i, &result);
        assert(ready == (i == 5));
    }
    assert(result.count == 1 && result.sum == 55);
    assert(result.minimum == 55 && result.maximum == 55);
    assert(result.first_scan == 5 && result.last_scan == 5);

    config = (struct adc_filter_config) {
        .input_div = 1, .osr = 4, .report_div = 1,
        .summary_mode = ADC_FILTER_SUMMARY_LATEST,
    };
    assert(!adc_filter_configure(&f, &config));
    assert(!adc_filter_set_postprocess(&f, 4, ADC_FILTER_ALPHA_ONE / 2));
    const uint16_t ewma_values[] = {
        100, 100, 100, 100, 200, 200, 200, 200, 100, 100, 100, 100,
    };
    const uint32_t ewma_expected[] = {100, 150, 125};
    for (uint32_t i = 0; i < 12; i++) {
        int ready = adc_filter_push_ex(&f, ewma_values[i], i, &result,
                                       &filtered_value, &filtered_ready);
        if ((i & 3) != 3) {
            assert(!ready && !filtered_ready);
            continue;
        }
        uint32_t group = i / 4;
        assert(ready && filtered_ready);
        assert(filtered_value == ewma_values[i]);
        assert(result.sum == ewma_expected[group]);
    }
    adc_filter_reset(&f, 1);
    for (uint32_t i = 0; i < 4; i++)
        assert(adc_filter_push(&f, 50, i, &result) == (i == 3));
    assert(result.sum == 50 && result.flags == ADC_FILTER_FLAG_DISCONTINUITY);

    config.osr = 0;
    assert(adc_filter_configure(&f, &config) < 0);
    (void)argc;
    (void)argv;
    puts("PASS: ADC filter decimation, rounding, summaries, and reset");
    return 0;
}

File: test/adc_filter_test.py
#!/usr/bin/env python3
import os
import pathlib
import random
import subprocess
import sys
import tempfile


ROOT = pathlib.Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))
from klippy.extras.adc_stream_model import FilterModel, run_interleaved


def test_c_reference_vectors():
    output = os.path.join(tempfile.gettempdir(), "adc_filter_test")
    subprocess.run([
        "cc", "-std=gnu11", "-Wall", "-Wextra", "-Werror",
        "-I", str(ROOT), "-I", str(ROOT / "src"),
        str(ROOT / "test" / "adc_filter_test.c"),
        str(ROOT / "src" / "generic" / "adc_filter.c"),
        "-o", output,
    ], check=True)
    result = subprocess.run(
        [output], check=True, capture_output=True, text=True)
    assert result.stdout.startswith("PASS:")


def independent_reference(samples, input_div, osr, shift, report_div):
    accepted = [(i, value) for i, value in enumerate(samples)
                if i % input_div == 0]
    outputs = []
    for offset in range(0, len(accepted) - osr + 1, osr):
        group = accepted[offset:offset + osr]
        total = sum(value for _, value in group)
        value = ((total + (1 << (shift - 1))) >> shift
                 if shift else total)
        outputs.append((group[-1][0], min(value, 0xffffffff)))
    reports = []
    for offset in range(0, len(outputs) - report_div + 1, report_div):
        group = outputs[offset:offset + report_div]
        reports.append({
            "count": report_div,
            "minimum": min(v for _, v in group),
            "maximum": max(v for _, v in group),
            "sum": sum(v for _, v in group),
            "first_scan": group[0][0],
            "last_scan": group[-1][0],
            "flags": 0,
        })
    return reports


def test_randomized_reference_model():
    rng = random.Random(0xADC2026)
    for _ in range(500):
        count = rng.randrange(1, 800)
        samples = [rng.randrange(4096) for _ in range(count)]
        input_div = rng.randrange(1, 9)
        osr = rng.randrange(1, 33)
        shift = rng.randrange(0, 9)
        report_div = rng.randrange(1, 17)
        model = FilterModel(input_div, osr, shift, report_div)
        actual = []
        for index, sample in enumerate(samples):
            result = model.push(sample, index)
            if result is not None:
                actual.append(result)
        expected = independent_reference(
            samples, input_div, osr, shift, report_div)
        assert actual == expected


def test_interleaved_channels_and_discontinuity():
    scans = [(i, 1000 - i) for i in range(32)]
    reports = run_interleaved(scans, [
        {"channel": 0, "filter": {
            "input_div": 1, "osr": 4, "shift": 2, "report_div": 2}},
        {"channel": 1, "filter": {
            "input_div": 2, "osr": 2, "shift": 1, "report_div": 2}},
    ])
    assert len(reports[0]) == 4
    assert len(reports[1]) == 4
    model = FilterModel(osr=2, shift=1)
    model.reset(discontinuity=True)
    assert model.push(1, 0) is None
    assert model.push(3, 1)["flags"] == 1


def test_window_average_and_ewma_reference():
    model = FilterModel(osr=4, window_divisor=4, alpha_q15=16384)
    reports = []
    for index, sample in enumerate([100] * 4 + [200] * 4 + [100] * 4):
        result = model.push(sample, index)
        if result is not None:
            reports.append(result["sum"])
    assert reports == [100, 150, 125]
    model.reset(discontinuity=True)
    result = None
    for index in range(4):
        result = model.push(50, index)
    assert result["sum"] == 50 and result["flags"] == 1


if __name__ == "__main__":
    test_c_reference_vectors()
    test_randomized_reference_model()
    test_interleaved_channels_and_discontinuity()
    test_window_average_and_ewma_reference()
    print("PASS: deterministic randomized ADC filter reference")
